feat: Implement Graph Pruning Defense and Evaluation Script

AlishThapa opened the pull request:
# 📦 Pull Request Template

## 📋 Summary

This pull request introduces a new defense mechanism, GraphPruningDefense, to the PyGIP framework.

This PR includes:

1.  GraphPruningDefense module, which is a compliant defense class, follows the IMPLEMENTATION.md
2.  Evaluation Script runs the defense on Cora, CiteSeer and PubMed datasets. If the model accuracy drops below 2%, then the optimal pruning ratio is determined and calculated.


## ✅ Checklist

- [x] My code follows the project's coding style
- [x] I have tested the changes and verified that they work
- [x] I have added necessary documentation (if applicable)
- [ ] I have linked related issues above (if any)
- [x] The PR is made from a feature branch, not `main`

---

## 🧠 Additional Context (Optional)
<img width="3600" height="2400" alt="pruning_effective_range_summary" src="https://github.com/user-attachments/assets/d3d2b7b1-651c-4297-a740-377da20cc767" />



<!--
PRs that do not follow this template may be delayed or closed. Thank you for contributing!
-->

AlishThapa committed:
feat: Implement Graph Pruning Defense and Evaluation Script

Edlison commented:
Could you please complete the implementation by adding the metrics from the original paper, namely TCA, ECA, TBA, EBA, and fidelity?

AlishThapa committed:
feat: implement full analysis for backdoor watermarking defense

AlishThapa committed:
implemented ECA and Fidelity in graph pruning defense, implemented evaluation of backdoor watermarking

AlishThapa committed:
implemented ECA and Fidelity in graph pruning defense, implemented evaluation of backdoor watermarking

AlishThapa committed:
Merge branch 'dev' into feat/graph-pruning-defense

AlishThapa commented:
I've completed the work on this PR. This includes the implementation of the `GraphPruningDefense` and a new, comprehensive evaluation for the existing BackdoorWM defense. Here’s a summary of the changes.

Graph Pruning Defense Enhancements
My initial task was to add the metrics **TCA**, **ECA**, **TBA**, **EBA**, and **Fidelity** to the GraphPruningDefense.

During development, I found that the TCA, TBA, and EBA are specific to backdoor-based defenses and they are not applicable to the graph pruning mechanism, which only removes edges.

Therefore, I have refactored the GraphPruningDefense and its evaluation script to focus on the metrics that are meaningful for this context:
- Test Accuracy (ECA): The primary performance metric for the pruned model.
- Fidelity: A new and valuable metric that measures how the model's predictive behavior changes as edges are removed.

The updated evaluation script (evaluate_graph_pruning.py) now generates an improved summary plot that visualizes both accuracy and fidelity, providing much richer insights into the effects of pruning.

**New: Backdoor Watermarking Evaluation**
I found that those metrices are applicable to the backdoor watermarking defense and the defense mechanism was already mentioned in the defense folder.
And to fulfill the original request of implementing all the specified metrics, I've added a new evaluation script for the pre-existing BackdoorWM.py defense.

A comprehensive summary plot that visualizes the final results across multiple datasets is also added.

In summary, this PR now delivers two distinct experiments: a refined analysis of Graph Pruning and a new, robust evaluation pipeline for Backdoor Watermarking.

Graph pruning data visualization.
<img width="4200" height="3600" alt="pruning_ECA_fidelity_summary" src="https://github.com/user-attachments/assets/d3de8544-8d02-41fd-9692-683f6b63b800" />

Backdoor watermarking data visualization.
<img width="2800" height="800" alt="full_defense_analysis_summary" src="https://github.com/user-attachments/assets/a0473407-87f9-4fa2-a428-f250bfb47bbc" />

AlishThapa committed:
removed comments, unwanted print statements and visualization function

AlishThapa committed:
removed comments, unwanted print statements and visualization function
